x-pack/metricbeat/module/mssql/fetcher.go

```diff
@@ -0,0 +1,150 @@
+// the database for each fetch operation
+func NewFetcher(uri string, qs []string, schema *s.Schema, log *logp.Logger) (*Fetcher, error) {
+	db, err := sql.Open("sqlserver", uri)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create db instance")
+	}
+
+	// Check the connection before executing all queries to reduce the number
+	// of connection errors that we might encounter.
+	if err = db.Ping(); err != nil {
+		return nil, err
+	}
+
+	f := &Fetcher{
+		queries: qs,
+		schema:  schema,
+		db:      db,
+		log:     log,
+	}
+
+	return f, nil
+}
+
+// Fetcher will make queries concurrently to the database to fetch results. It
+// must be created by each metricset and fed with the queries that must be
+// executed, a SQL implementor and a pointer to the schema to "apply". A
+// metricset could, potentially, need more than one query to fill all its
+// results.
+type Fetcher struct {
+	schema *s.Schema // Schema is the metricset schema to apply to the fetched data to turn it into a result mapstr.
+	db     *sql.DB   // Database to execute queries against.
+	log    *logp.Logger
+
+	queries []string // List of queries to execute concurrently.
+}
+
+// Close closes db connection to the server
+func (f *Fetcher) Close() error {
+	return f.db.Close()
+}
+
+// Report receives a mb.ReporterV2 to send the data to the outputs. Because the operations are common between metricsets
+// it can be extracted to a common function
+func (f *Fetcher) Report(reporter mb.ReporterV2) {
+	var err error
+	for _, q := range f.queries {
+		if err = f.getEventsWithQuery(f.db, q, reporter); err != nil {
+			logp.Error(err)
+		}
+	}
+}
+
+type rowsResultHandler struct {
+	dest        []interface{}
+	rows        *sql.Rows
+	reporter    mb.ReporterV2
+	columnNames []string
+}
+
+func newRowsResult(reporter mb.ReporterV2, columnNames []string, rows *sql.Rows) *rowsResultHandler {
+	return &rowsResultHandler{
+		dest:        make([]interface{}, len(columnNames)),
+		columnNames: columnNames,
+		rows:        rows,
+		reporter:    reporter,
+	}
+}
+
+// getEventsWithQuery performs the query on the database and converts the rows
+// to an slice of common.MapStr.
+func (f *Fetcher) getEventsWithQuery(db *sql.DB, query string, reporter mb.ReporterV2) error {
+	// Returns the global status, also for versions previous 5.0.2
+	rows, err := db.Query(query)
+	if err != nil {
+		return errors.Wrapf(err, "error performing db query='%v'", query)
+	}
+	defer func() {
+		if err = rows.Close(); err != nil {
+			logp.Error(errors.Wrap(err, "error closing query row"))
+		}
+	}()
+
+	columnNames, err := rows.Columns()
+	if err != nil {
+		return errors.Wrap(err, "error getting column names")
+	}
+
+	rowsResult := newRowsResult(reporter, columnNames, rows)
+
+	if err = rowsResult.handle(f.schema); err != nil {
+		return errors.Wrap(err, "could not convert rows result to beats mapstr")
+	}
```

Comment:
When I said to pass the row as argument and to keep all the iteration in the same place, I meant to do something like:
```
row := make(...)
for rows.Next() {
    if err = rows.Scan(row...); err != nil {
        ...
    }
    f.handleRow(reporter, columnNames, row)
}
```

Reply:
I'd like more the current change. You were right, there was a mix of responsibilities that makes the code a bit confusing. `Fetcher` which is a struct that _must fetch data from the database` was iteraring over the row results to get individual rows. Now it just calls `rowsResult.handle` so that the responsibility of extracting data from rows is fully abstracted from `Fetcher` into the `rowsResultHandler` which, in fact, could be in its own file even.

Reply:
I was not refering to separating responsibilities, but good point, actually this can be important in case we reuse this for other sql services.

The only thing now is that I see this handler struct a bit artificial:
* I'd expect for a handler to have a method like `handle(the thing to be handled)`
* I wouldn't expect the handler to keep state, why is `dest` a member of the struct and not just a local variable in `handle`? (well, this can be seen as some kind of optimization as the object also knows about the columns)
* I don't see a clear separation between the members of the handler struct and the inputs of the handle method. 
for example if handling here means _to apply an schema to generate an event and report it_ why does it contain the reporter, but not the schema?

I won't block this PR because of these questions, it can be a matter of code styles. And in any case this has to be revisited when we move this fetcher to be a reusable helper.

Reply:
I'll write down this comments to have a look in a following PR. I understand your point. Thanks Jaime :)